src/shared/machine-credential.c

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "creds-util.h"
+#include "escape.h"
+#include "extract-word.h"
+#include "fileio.h"
+#include "macro.h"
+#include "memory-util.h"
+#include "machine-credential.h"
+#include "path-util.h"
+#include "string-util-fundamental.h"
+
+static void machine_credential_done(MachineCredential *cred) {
+        assert(cred);
+
+        cred->id = mfree(cred->id);
+        cred->data = erase_and_free(cred->data);
+        cred->size = 0;
+}
+
+void machine_credential_free_all(MachineCredential *creds, size_t n) {
+        assert(creds || n == 0);
+
+        FOREACH_ARRAY(cred, creds, n)
+                machine_credential_done(cred);
+
+        free(creds);
+}
+
+int machine_credential_set(MachineCredential **credentials, size_t *n_credentials, const char *cred_string) {
+        _cleanup_free_ char *word = NULL, *data = NULL;
+        MachineCredential *creds = *ASSERT_PTR(credentials);
+        ssize_t l;
+        size_t n_creds = *ASSERT_PTR(n_credentials);
+        int r;
+        const char *p = ASSERT_PTR(cred_string);
+
+        if (n_creds > 0)
+                ASSERT_PTR(creds);
+
+        r = extract_first_word(&p, &word, ":", EXTRACT_DONT_COALESCE_SEPARATORS);
+        if (r == -ENOMEM)
+                return r;
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse --set-credential= parameter: %m");
```

Comment:
hmm, machine_credential_set()'s primary focus appears to be to parse cmdline params, and hence is strictly user-facing. Hence I think this might actually be a good function to do logging above LOG_DEBUG, i.e. log_error_errno()

I mean, the errors you output here are really useful to people, we shuld show them.

In systemd we want that functions either are of the logging kind, or not of the logging kind. (The latter means they only log at LOG_DEBUG level). While most of the stuff in src/basic/ + src/shared/ is probably of the not-logging kind, this isn't strictly so: it's OK if something is clealy user-facing to immediately log.

Or in other words, I think this should feel a bit more like the stuff in parse-argument.[ch], whose only purpose too is to just parse command line arguments, and immediately complain.

Reply:
That makes sense, thanks for the explanation :)